Allow clients to subscribe to slot migrations

zuiderkwast opened the pull request:
The cluster bus doesn't distinguish between migrations and other actions.
This commit detects a migration by checking if exactly one slot has a new
owner.

Only one moved slot per cluster bus message or command is notified. This
is to avoid flooding the clients. In other cases, such as failovers,
clients will need to handle MOVED redirects and update the slot mapping.

A special pubsub channel `__redis__:moved` is used (naming style as in
client-side caching). The message is a string on the form "MOVED slot
endpoint:port" very similar to MOVED redirects.

Some special logic added to avoid creating the pubsub message if there
are no subscribers.

Fixes #10150.

PingXie commented on src/cluster.c:
```diff
@@ -596,6 +601,9 @@ void clusterInit(void) {
     server.cluster->todo_before_sleep = 0;
+    server.cluster->moved_slot_since_sleep = MOVED_SLOT_NONE;
+    server.cluster->moved_slot_channel =
+        createObject(OBJ_STRING, sdsnew("__redis__:moved"));
```
This is the second "system" channel. So far it looks like this "system" property is implied only and not explicitly enforced by Redis, meaning that any client could publish to a channel whose name starts with the "\_\_redis\_\_:" prefix. I also didn't find the "\_\_redis\_\_:" prefix mentioned in the documentation.

I wonder if we should consider defining a clear contract and enforcing it in Redis. For starters, would the below make sense?

1. Client can't publish to any channel whose name starts with "\_\_redis\_\_:"
2. Client subscribing to a channel whose name starts with "\_\_redis\_\_:" won't result in the channel getting created silently. Instead, Redis would fail the subscription if such a "system" channel doesn't already exist. This contract helps with "system" channel discovery.

zuiderkwast replied:
We have these system channels already:

* `__redis__:invalidate` is used in client-side caching.
* `__keyspace@0__:*` and `__keyevent@0__:*` used in keyspace notifications. These are dynamic and contain the db number (0) in the prefix. After the prefix is a key or a command.

I don't mind restricting channels with the `__redis__:` prefix but I don't think it's required either. It can be done later. It's a breaking change but probably an acceptable one.

PingXie replied:
Agreed it is an orthogonal change that can be done separately. Timing wise though, it would be nice if we could introduce this breaking change (tightening up system channels) as part 7.0 release. I will file a bug to track the discussion independently.

PingXie commented on src/cluster.c:
```diff
@@ -4147,17 +4155,49 @@ void clusterCron(void) {
+        int use_pport = (server.tls_cluster && n->pport &&
+                         c->conn && connGetType(c->conn) != CONN_TYPE_TLS);
+        if (messages[use_pport] == NULL) {
+            sds s = sdscatprintf(sdsempty(),
```
I like the idea of reusing the MOVED message but can we make it more explicit by single-sourcing the string formatting logic? Or at least add some comments to remind the next person inline that the string generated here should match that of `clusterRedirectClient`?

zuiderkwast replied:
Of course, I'll single-source it.

zuiderkwast replied:
Done

PingXie commented on src/cluster.c:
```diff
@@ -4177,6 +4217,16 @@ void clusterBeforeSleep(void) {
 
+    /* Notify clients subscribed to moved slot events. To avoid flooding the
+     * clients, we only publish the moved slot message if exactly one slot has
+     * been migrated. In cases like failover, clients will receive -MOVED
```
Can we consider sending the MOVED_SLOT_MULTIPLE (-2) message anyways? This would give clients a chance to refresh the cluster topology outside the data serving path while not adding too much overhead.

zuiderkwast replied:
We could. We could also notify clients about changed replicas (no slots moved). Mostly we need to decide if we want to send it in the same channel or a different one and the format of the message. WDYT?

PingXie replied:
Separate channels sound cleaner to me as it would allow easy filtering. For instance, I wouldn't imagine all clients care about the addition or removal of replicas.

PingXie commented on src/cluster.c:
```diff
@@ -4155,30 +4155,38 @@ void clusterCron(void) {
+        connGetType(c->conn) != CONN_TYPE_TLS;
+}
+
+/* For redirects, verb must start with a dash, e.g. "-ASK" or "-MOVED". */
```
Maybe consider converting code comments into strong typed arguments? One idea could be an enum class with two values: ASK = 1 and MOVED = 2.  Then we could assert on invalid verbs trivially.  Extending the idea further, the function could take another argument to indicate whether the string created is intended to be used as a normal error string or a pubsub message.

zuiderkwast replied:
I did it first, but we also need with dash for redirects (resp errors) and without dash for pubsub. Flags maybe? I think a string like this is at least simple.

PingXie replied:
Flags were what I thought as well. That said, I don't have a strong opinion here since I don't see the use case of this function getting expanded further and we have tests to validate the response/message strings.

PingXie commented on src/cluster.c:
```diff
@@ -4224,7 +4232,7 @@ void clusterBeforeSleep(void) {
      * including replicas. */
     if (moved_slot >= 0) {
```
remove the `if (moved_slots >= 0)` check? I thought we agreed on sending the special multiple-slots-moved message?

zuiderkwast replied:
No we discussed how it should be sent and if it should be in a separate channel. (We can't send "MOVED -2 ip:port". If I just remove the if, the code will crash.)

PingXie replied:
> If I just remove the if, the code will crash

Curious - was it Redis that crashed or the test? Any challenge/risk in fixing this crash?

> it should be in a separate channel

That sounds a bit counter-intuitive to me, if I as a client need to subscribe to two different channels to get the same type of information. The difference between the two channels sounds more like an implementation detail than something conceptual. I was under the impression that the separate channel would be used for different types of events, such as the arrival/departure of a replica.

zuiderkwast replied:
> > If I just remove the if, the code will crash
> 
> Curious - was it Redis that crashed or the test? Any challenge/risk in fixing this crash?

clusterNotifyMovedSlot accesses `server.cluster->slots[moved_slot]`. With a negative index, it's read outside the allocation. Without the  `if >= 0`, it'd also try with -1 when there is no slot moved. I must assume you misunderstand. It needs more implementation if we want to notify for multiple slots and in particular, it needs to be decided what the notification should look like.

> > it should be in a separate channel
> 
> That sounds a bit counter-intuitive to me, if I as a client need to subscribe to two different channels to get the same type of information. The difference between the two channels sounds more like an implementation detail than something conceptual. I was under the impression that the separate channel would be used for different types of events, such as the arrival/departure of a replica.

I thought you preferred a separate channel in this comment where we talked about multiple moved and also other events: https://github.com/redis/redis/pull/10358#discussion_r818893065

zuiderkwast replied:
So, exactly what behaviour do you suggest for multiple moved slots? This often means failover but the cluster bus doesn't specify that.

PingXie replied:
I misunderstood your [comment](https://github.com/redis/redis/pull/10358#discussion_r818426335). I thought it was meant for different types of events. 

I think there should be one channel for each type of messages and the type discussed here is "slot ownership". The reason why the slot ownership changes isn't that important to the client (be it slot migration or failover) IMO. The use case I have in mind is to facilitate faster/cheaper routing information update.

zuiderkwast replied:
OK, now we sorted out the misunderstandings. :-)

In the case of failover, we might have something like 5000 slots moved at once. Should we send 5000 messages to the subscribers in that case? It is flooding the clients with too much data if we do it in the naive way. We need to send a more compact message which includes a list of slot ranges.

.... or we can just send a message like "RELOAD" (or "INVALIDATE" or "RESHARDED") which tells the client to reload the slot mapping using CLUSTER SLOTS, SHARDS or NODES, whatever the client prefers.

... or we can send no message at all, which will result in a -MOVED redirect sooner or later, which triggers the client to reload the slot mapping. This is my initial choice in this PR, because it is minimalistic and extensible.

yossigo commented:
@zuiderkwast It is not explicitly stated, but I assume the main goals here are:
* Avoid potential latency spike with lazy processing of `-MOVED` (as a redirect, or as a trigger for `CLUSTER SLOTS`)
* Optimize clients that maintain a connection pool and can use a single connection for this kind of signalling.

Am I missing something else here?
Regarding additional messages or channels, I think we should define it now even if the implementation is partial because it takes time for clients to catch up.

zuiderkwast commented:
@yossigo That's right. Not necessarily a connection pool, but long-running connections, yes.

Clients that reload the complete slot mapping on -MOVED redirects will not have to do that if only a single slot has been moved. Slot migration is one slot at a time. When moving a lot of slots, the slot mapping is updated one slot at a time and without this feature clients would need to reload the full slot mapping many times.

If we decide how to notify about *multiple slots moved* and other changes, I can as well implement it now.

madolson commented on src/cluster.c:
```diff
@@ -596,6 +601,9 @@ void clusterInit(void) {
     server.cluster->todo_before_sleep = 0;
+    server.cluster->moved_slot_since_sleep = MOVED_SLOT_NONE;
+    server.cluster->moved_slot_channel =
+        createObject(OBJ_STRING, sdsnew("__redis__:moved"));
```
I would suggest something like, `__redis__:cluster_moved`. We might want to include other cluster notifications like failovers and state changes (like preferred endpoints) in the future, and they should probably be prefixed the same way. (Or perhaps we could also have something like`__redis__:cluster_all` that includes all notifications as well?

PingXie replied:
how about ~~`cluster_slot_moved`~~ `cluster_slots_moved`?

madolson commented on src/cluster.c:
```diff
@@ -4147,17 +4155,57 @@ void clusterCron(void) {
+        connGetType(c->conn) != CONN_TYPE_TLS;
+}
+
+/* For redirects, verb must start with a dash, e.g. "-ASK" or "-MOVED". */
```
We might consider just adding the dash from the calling site instead of documenting it here. The dash is to indicate that it's an error response. (We don't actually even need it I think, since we should append - on errors regardless)

zuiderkwast replied:
It is added in the calling site. I can delete this comment if that's what you mean. (If we don't add id, "-ERR " is prepended by the addErrorReply functions.)

madolson reviewed (commented):
I agree with @yossigo that I think we need a broader idea of all the types of topology messages we want to send, so we plan for it.

PingXie commented:
+1 on defining a complete message type system. As far as the moved channel is concerned what specific information do you have in mind @madolson? If we agree on the use case of facilitating the routing information update, mimicking -ASK/-MOVED makes sense to me.

zuiderkwast commented:
> I agree with @yossigo that I think we need a broader idea of all the types of topology messages we want to send, so we plan for it.

@madolson Thx for reviewing. Sounds good. In the cluster bus, a node sends its full info, so if we want to send partial info to the client, we would have to compute some kind of diff. Perhaps it's enough to send the full state for a node. The same with slots. If we don't send all slots for a node, we'd need to compute a diff to find out which slots are moved. (Detecting a single moved slot is simple though.)

The possible events I'm aware of are failover, replica migration, added node, removed node, state change and slot migration.

We could structure a message like a partial CLUSTER SHARDS reply but it would become quite large in some cases, unless we detect events like replica migration (master id changed and no other info changed). The events I think we'd need to detect are failover, replica migration, added node, removed node, state change, slot migration.

> The implementation feels a bit off to me, in that we are sending "moved redirect" messages to clients. It seems like we should be able to send something richer with more information.

I agree. The three reasons I picked the redirects format are 1) pubsub messages are always strings so far, 2) clients are familiar with redirects and already have code for parsing them and 3) the message is very compact.

We can of course introduce a RESP-structured pubsub message here.

zuiderkwast commented:
Coming back to this one.

The reason I want to subscribe specifically to the single-slot-migration scenario: When a bunch of slots are migrated, this is done one slot at a time and it can be quite slow. If there are some very active clients, they will run into MOVED redirect after each migrated slot, which typically causes them to reload the slots map. When the next slot is migrated, they will run into another MOVED redirect, and so on.

If CLUSTER SLOTS is O(N), the total work done by each client will be O(N*S) where S is the number of slots being moved, if they call CLUSTER SLOTS on each MOVED redirect, after each migrated slot.

All the other topology changes (failover, etc.) normally cause the clients to reload the slots mapping only once. In this case, waiting for a redirect before reloading the slot mapping is totally fine IMO.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/redis/redis?pullRequest=10358) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/redis/redis?pullRequest=10358) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/redis/redis?pullRequest=10358) it.</sub>
